### PR TITLE
Adding: A Video Vision Transformer

### PR DESCRIPTION
**Pull request by ariG23498**
In this example we (with [Ayush Thakur](https://github.com/ayulockin)) have minimally implemented [ViViT](https://arxiv.org/abs/2103.15691). Here we talk about a new tokenization scheme (tubelet embedding) that tokenizes videos. This embedding scheme not only encodes the frame pathces but also helps encode the temporal componenet of the video. With this embedding we train the tokens on a simple vision transformer.

Colab: [ViViT](https://colab.research.google.com/drive/1VK06kg8125BlfYd17-fC7LjrP0myksIl?usp=sharing)

### CODE REVIEW

File: examples/vision/vivit.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
+"""
+Title: Video Vision Transformer
+Author: [Aritra Roy Gosthipaty](https://twitter.com/ariG23498), [Ayush Thakur](https://twitter.com/ayushthakur0) (equal contribution)
+Date created: 2022/01/12
+Last modified:  2022/01/12
+Description: A pure transformer based architecture for video classification
+"""
+"""
+<a href="https://colab.research.google.com/github/ariG23498/ViViT/blob/ayut/ViViT.ipynb"
```
Comment (fchollet) [inline examples/vision/vivit.py:9]: Please remove this link, but you can copy / paste it in the PR description

File: examples/vision/vivit.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+"""
+## Imports
+"""
+
+import os
+import io
+import imageio
+import numpy as np
+import matplotlib.pyplot as plt
+from medmnist import INFO
```
Comment (fchollet) [inline examples/vision/vivit.py:54]: Just import `medmnist` and access its attributes directly

File: examples/vision/vivit.py [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+from IPython.display import Image, display
+from ipywidgets import widgets, HTML
```
Comment (fchollet) [inline examples/vision/vivit.py:59]: Likewise

File: examples/vision/vivit.py [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+
+"""
+### `tf.data` pipeline
+"""
+
+
+@tf.function
+def preprocess(frames: tf.Tensor, label: tf.Tensor):
+    """
+    Preprocess the frames tensors and parse the labels
```
Comment (fchollet) [inline examples/vision/vivit.py:151]: Move sentence one line above

File: examples/vision/vivit.py [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+
+
+def prepare_dataloader(
+    videos: np.ndarray,
+    labels: np.ndarray,
+    loader_type: str = "train",
+    batch_size: int = BATCH_SIZE,
+):
+    """
+    Utility function to prepare dataloader.
```
Comment (fchollet) [inline examples/vision/vivit.py:173]: Same

File: examples/vision/vivit.py [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+    """
+    dataset = tf.data.Dataset.from_tensor_slices((videos, labels))
+
+    if loader_type == "train":
+        dataset = dataset.shuffle(1024)
```
Comment (fchollet) [inline examples/vision/vivit.py:178]: Prefer making buffer size dependent on the batch size

File: examples/vision/vivit.py [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+
+| ![tubelet embedding](https://i.imgur.com/9G7QTfV.png) |
+| :--: |
+| Tubelet Embedding [Source](https://arxiv.org/abs/2103.15691) |
+"""
+
+
+class TubeletEmbedding(layers.Layer):
+    def __init__(self, embed_dim=PROJECTION_DIM, patch_size=PATCH_SIZE):
+        super().__init__()
```
Comment (fchollet) [inline examples/vision/vivit.py:219]: Add `**kwargs` support

File: examples/vision/vivit.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+are then flattened and projected to build video tokens.
+
+| ![tubelet embedding](https://i.imgur.com/9G7QTfV.png) |
+| :--: |
+| Tubelet Embedding [Source](https://arxiv.org/abs/2103.15691) |
+"""
+
+
+class TubeletEmbedding(layers.Layer):
+    def __init__(self, embed_dim=PROJECTION_DIM, patch_size=PATCH_SIZE):
```
Comment (fchollet) [inline examples/vision/vivit.py:218]: Make these positional args (no default) and pass the hardcoded value upon instantiation -- this makes the layer easier to reuse elsewhere

File: examples/vision/vivit.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+
+"""
+## Positional Embedding
+
+This layer adds positional information to encoded video tokens.
+"""
+
+
+class PositionalEncoder(layers.Layer):
+    def __init__(self, embed_dim=PROJECTION_DIM):
```
Comment (fchollet) [inline examples/vision/vivit.py:242]: Likewise

File: examples/vision/vivit.py [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+        super().__init__()
```
Comment (fchollet) [inline examples/vision/vivit.py:243]: Likewise

File: examples/vision/vivit.py [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+
+    def build(self, input_shape):
+        _, num_tokens, _ = input_shape
+        self.position_embedding = layers.Embedding(
+            input_dim=num_tokens, output_dim=self.embed_dim
+        )
+        self.positions = tf.range(start=0, limit=num_tokens, delta=1)
+
+    def call(self, encoded_tokens):
+        # encode the positions and add it to the encoded tokens
```
Comment (fchollet) [inline examples/vision/vivit.py:254]: Capitalize first character of all comments

File: examples/vision/vivit.py [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+
+        # Skip connection
+        encoded_patches = layers.Add()([x3, x2])
+
+    # Layer normalization and Global average pooling.
+    representation = layers.LayerNormalization(epsilon=LAYER_NORM_EPS)(encoded_patches)
+    representation = layers.GlobalAvgPool1D()(representation)
+
+    # Classify outputs.
+    outputs = layers.Dense(NUM_CLASSES, activation="softmax")(representation)
```
Comment (fchollet) [inline examples/vision/vivit.py:319]: num_classes should be an argument of this function

File: examples/vision/vivit.py [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+                layers.Dense(units=embed_dim, activation=tf.nn.gelu),
+                layers.Dense(units=embed_dim, activation=tf.nn.gelu),
+            ]
+        )(x3)
+
+        # Skip connection
+        encoded_patches = layers.Add()([x3, x2])
+
+    # Layer normalization and Global average pooling.
+    representation = layers.LayerNormalization(epsilon=LAYER_NORM_EPS)(encoded_patches)
```
Comment (fchollet) [inline examples/vision/vivit.py:315]: Should be an argument of this function

File: examples/vision/vivit.py [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+def create_vivit_classifier(
+    tubelet_embedder,
+    positional_encoder,
+    transformer_layers=NUM_LAYERS,
+    num_heads=NUM_HEADS,
+    embed_dim=PROJECTION_DIM,
+):
+
+    # Get the input layer
+    inputs = layers.Input(shape=INPUT_SHAPE)
```
Comment (fchollet) [inline examples/vision/vivit.py:285]: Should be an argument of this function

File: examples/vision/vivit.py [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+    boxes.append(make_box_for_grid(ib, caption))
+
+widgets.GridBox(boxes, layout=widgets.Layout(grid_template_columns="repeat(5, 200px)"))
+
+"""
+## Final Thoughts
+
+With a vanilla implementation we achieve ~75-77% Top-1 accuracy on the test dataset. 
+
+Places to improve:
```
Comment (fchollet) [inline examples/vision/vivit.py:430]: Make a full sentence

**Review by fchollet — COMMENTED** [review]
Awesome example, thank you for the PR!

**Comment by ariG23498** [comment]
@fchollet thanks for the detailed review. We have made the necessary changes.

**Review by fchollet — APPROVED** [review]
Thanks for the update! I pushed some copyedits, please pull them first. You can then add the generated files. Thanks for the great contribution! 👍

**Comment by ariG23498** [comment]
@fchollet the generated files have been added.

**Comment by fchollet** [comment]
Thank you! I'll take it from here.

**Pull request merged**